iocore/utils/SharedExtendible.h

```diff
@@ -0,0 +1,447 @@
+    {
+      auto field_iter = schema.fields.find(field_name);
+      ink_release_assert(field_iter != schema.fields.end());
+      FieldSchema &fs = field_iter->second;
+      ink_release_assert(fs.access == FieldAccess_e);
+      return FieldId_t(fs.offset);
+    }
+
+    // only allow offest to be used by schema code
+    friend SharedExtendible_t;
+
+  private:
+    FieldId(SharedOffset_t const &offset) { offset_ptr = &offset; }
+    SharedOffset_t
+    offset() const
+    {
+      return *offset_ptr;
+    }
+  };
+  using BitFieldId = FieldId<BIT, bool>;
+
+  /////////////////////////////////////////////////////////////////////
+  /// defines a runtime "member variable", element of the blob
+  struct FieldSchema {
+    using Func_t = void(void *);
+
+    SharedAccessEnum access; ///< which API is used to access the data
+    SharedOffset_t size;     ///< size of field
+    SharedOffset_t offset;   ///< offset of field from 'this'
+    Func_t *construct_fn;    ///< the data type's constructor
+    Func_t *destruct_fn;     ///< the data type's destructor
+  };
+
+  /////////////////////////////////////////////////////////////////////
+  /// manages the fields as data structures
+  class Schema
+  {
+    friend SharedExtendible_t;
+
+  private:
+    std::unordered_map<std::string, FieldSchema> fields; ///< defined elements of the blob by name
+    uint32_t bit_offset;                                 ///< offset to first bit
+    size_t alloc_size;                                   ///< bytes to allocate
+    size_t alloc_align;                                  ///< alignment for each allocation
+    std::atomic_uint instance_count;                     ///< the number of class instances in use.
+
+  public:
+    /// Constructor
+    Schema() : instance_count(0) { reset(); }
+
+    /// Add a new Field to this record type
+    template <SharedAccessEnum Access_t, typename Field_t>
+    bool
+    addField(FieldId<Access_t, Field_t> &field_id, std::string const &field_name)
+    {
+      static_assert(Access_t == BIT || std::is_same<Field_t, bool>::value == false,
+                    "Use BitField so we can pack bits, they are still atomic.");
+      ink_release_assert(instance_count == 0); // it's too late, we already started allocating.
+
+      SharedOffset_t size = 0;
+      switch (Access_t) {
+      case BIT: {
+        size = 0;
+      } break;
+      case ATOMIC: {
+        size        = std::max(sizeof(std::atomic<Field_t>), alignof(std::atomic<Field_t>));
+        alloc_align = std::max(alloc_align, alignof(std::atomic<Field_t>));
+      } break;
+      default:
+        size = sizeof(Field_t);
+      }
+
+      // capture the default constructors of the data type
+      static auto construct_fn = [](void *ptr) { new (ptr) Field_t; };
+      static auto destruct_fn  = [](void *ptr) { static_cast<Field_t *>(ptr)->~Field_t(); };
+
+      fields[field_name]  = FieldSchema{Access_t, size, 0, construct_fn, destruct_fn};
```

Comment:
Why bother with `static` locals, instead of passing the lambda's directly?

Reply:
My thought was that if it is the same type, then use the same lambda. No need to instance a new lambda. Maybe the compiler is smart enough to do that anyways.